### ISSUE TITLE
WARNING on merging the same dict data

### ISSUE THREAD

**Issue by yeled** [leapfrogonline/ansible-merge-vars#22]
**Describe the Bug**
I get warning when the data in the {} is exactly the same. Even empty. This is the simplest bug report I could share.

I am using `ansible-merge-vars==3.0.0`, I tried `4.0.0` but it didn't change anything.

What am I doing wrong?

```
TASK [routing : Merge Aggregates] ****************************************************************************************************************************
task path: /Users/yeled/src/gh/network/ansible/roles/routing/tasks/main.yml:1
Friday 23 November 2018  12:52:57 +0000 (0:00:00.518)       0:00:10.563 *******
 [WARNING]: aggregates is already defined, are you sure you want to overwrite it?
The contents of aggregates are: {}
Merging vars in this order: []
ok: [leaf-1-lab.net.cp1-iad.github.net] => {
    "ansible_facts": {
        "aggregates": {}
    },
    "ansible_facts_cacheable": false,
    "changed": false
}
```

**Ansible Playbook**

```yaml
- name: Merge Aggregates
  merge_vars:
    suffix_to_merge: __aggregates__to_merge
    merged_var_name: aggregates
    expected_type: 'dict'
    recursive_dict_merge: yes
  when:
    - not unmerged_aggregates is defined


- name: "Routing Aggregates"
  template: src={{os}}/aggregates.j2 dest=../../tmp/fragments/{{inventory_hostname}}/aggregates.conf
  when:
    - aggregates is defined
    - inventory_hostname_short is match("bdr.*|!ssw.*|!spine.*")

- name: "Default VRF route"
  template: src={{os}}/default-vrf.j2 dest=../../tmp/fragments/{{inventory_hostname}}/iproute-default-vrf.conf
  when:
    - inventory_hostname_short is match("^leaf.*")
```

**Environment**
```
ansible 2.7.0
  config file = /Users/yeled/src/gh/network/ansible/etc/ansible.cfg
  configured module search path = [u'/Users/yeled/src/gh/network/ansible/env/lib/python2.7/site-packages/napalm_ansible']
  ansible python module location = /Users/yeled/src/gh/network/ansible/env/lib/python2.7/site-packages/ansible
  executable location = /Users/yeled/src/gh/network/ansible/env/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

**Comment by ibarrere** [comment]
I know you posted this ages ago, but I see the same thing. It looks like the variable gets cached, so the first time you run it for awhile it shouldn't give you this warning, but then if you run it consecutively you will see it since the var is already cached.

**Comment by yeled** [comment]
that is exactly the issue. We ended up re-writing a simpler version that doesn't automagically find the vars, but does make sure they aren't cached. Works for us.

**Comment by fajpunk** [comment]
Sorry for the super late response on this, thanks for adding more details!

Can you describe in more detail what you mean when you say the variable gets cached?  Do you mean cached within the same Ansible run, or in between Ansible runs? I'm curious what you did in your version to make sure that the variables don't get cached, @yeled 

Similarly, @ibarrere, when you say you see the problem "if you run it consecutively", do you mean if the same `merge_vars` task gets run twice in the same play?  Maybe if the `merge_vars` task is defined in a role and that role gets executed twice, either with different parameters or on different hosts?

If this is the issue, then one solution I'm thinking of is to only show the warning if the existing and the new values for the merged variable are different.

For example, I see that warning on this playbook:
```
- name: Running merge_vars twice
  hosts: localhost
  gather_facts: false  # Speed up the example
  vars: # Note that these could be definied anywhere in inventory
    group1_ports__to_merge:
      - 22
      - 80
    group2_ports__to_merge:
      - 22
      - 1111
  tasks:
    - name: Merge port vars
      merge_vars:
        suffix_to_merge: ports__to_merge
        merged_var_name: merged_ports
        expected_type: list

    - name: Merge port vars
      merge_vars:
        suffix_to_merge: ports__to_merge
        merged_var_name: merged_ports
        dedup: false
        expected_type: list

    - debug:
        var: merged_ports
```
```
$ ansible-playbook -v examples/merge_lists_no_dedup_playbook.yml
No config file found; using defaults
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [Running merge_vars twice] **********************************************************************************************************************************************

TASK [Merge port vars] *******************************************************************************************************************************************************
Merging vars in this order: ['group1_ports__to_merge', 'group2_ports__to_merge']
ok: [localhost] => {"ansible_facts": {"merged_ports": [22, 80, 22, 1111]}, "ansible_facts_cacheable": false, "changed": false}

TASK [Merge port vars] *******************************************************************************************************************************************************
 [WARNING]: merged_ports is already defined, are you sure you want to overwrite it?

The contents of merged_ports are: [22, 80, 22, 1111]
Merging vars in this order: ['group1_ports__to_merge', 'group2_ports__to_merge']
ok: [localhost] => {"ansible_facts": {"merged_ports": [22, 80, 22, 1111]}, "ansible_facts_cacheable": false, "changed": false}

TASK [debug] *****************************************************************************************************************************************************************
ok: [localhost] => {
    "merged_ports": [
        22,
        80,
        22,
        1111
    ]
}

PLAY RECAP *******************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

**Comment by ibarrere** [comment]
@fajpunk 

I'm referring to between ansible runs. So it appears that the merged variable gets cached with the host_vars just like all the other host_vars, but then times out eventually. So, if I log on and run a play that involves merge_vars, it won't give me any warnings since it's the first run of the day. If I run it again shortly after that (and any time up until the host_vars time out and disappear) then I get the warning. I haven't tried a play like yours above where you explicitly run merge_vars twice, in my play I just run it once.

I thought that the cacheable option was meant to control this behavior (and maybe it is), but it doesn't seem to make any difference, as I don't have it set on the play I'm working with. In any case, it's not a huge issue as much as a minor annoyance, so changing the warning behavior to only warn if the overwriting values differ from the cached values is probably fine.

**Comment by yeled** [comment]
@ibarrere has it right. I don't use the leapfrogonline version anymore but their explanation is spot on.

We just use a Jinja filter now, instead of an action. In that way it is never cached.

**Comment by fajpunk** [comment]
Thanks to both of you for the responses!

This is a wacky one.  I believe the 'cacheable' arg to this plugin broke between Ansible 2.4 and 2.5 with this change:
https://github.com/ansible/ansible/pull/32302/files#diff-b893b8b8319cd4518208dd4fe53f3d7bR516

With Ansible from that change on, _any_ variable that is set by a module or action plugin other than `set_fact` will be cached if fact caching is enabled, regardless of whether or not `ansible_facts_cacheable` (or `_ansible_facts_cacheable`) is set in the return value from the `run`method.  It doesn't seem like there is any way to set a non-persistent fact from a custom action plugin.

Given that, I'm inclined to just remove the warning (and the `cacheable` option).  The original purpose of the warning was to alert the user to possibly picking a name for their merged var that was already explicitly defined elsewhere in the inventory, but it looks like it can't serve that purpose.

I'll think about it a bit more and do a bit more digging, but there will likely be a release coming soon _sans_ warning :)

**Comment by fajpunk** [comment]
Done and released on pypi as 5.0.0.  That warning should be gone for you with that version, @ibarrere.  Feel free to re-open this issue if you're still running into it!
